runtime/compiler/z/codegen/S390PrivateLinkage.cpp

```diff
@@ -1526,6 +1526,246 @@ J9::Z::PrivateLinkage::createEpilogue(TR::Instruction * cursor)
+   int32_t numImplementers = chTable->findnInterfaceImplementers(declaringClass, MAX_IMPLEMENTERS_TO_EVALUATE, implArray, cpIndex, methodSymRef->getOwningMethod(comp), comp);
+   if ((numImplementers != 0) && (numImplementers <= MAX_IMPLEMENTERS_TO_EVALUATE))
+      {
+      // Find out how many interfaces each implementer implements
+      int32_t maxInterfaces = MIN_ITABLE_ITERATIONS;
+
+      for (int32_t i=0; i < numImplementers; ++i)
+         {
+         TR_OpaqueClassBlock *containingClass = implArray[i]->containingClass();
+         int32_t numInterfaces = fej9->numInterfacesImplemented((J9Class *)containingClass);
+         maxInterfaces = (numInterfaces > maxInterfaces) ? numInterfaces : maxInterfaces;
+         }
+
+      iterations = (maxInterfaces > MAX_ITABLE_ITERATIONS) ? MAX_ITABLE_ITERATIONS : maxInterfaces;
+
+      if (trace)
+         traceMsg(comp, "ITable declaringClass %p numImplementers %d maxInterfaces %d iterations %d\n", declaringClass, numImplementers, maxInterfaces, iterations);
+      }
+   else if (trace)
+      traceMsg(comp, "ITable iteration is the default value of %d.\n", iterations);
+   return iterations;
+   }
+
+TR::Instruction *
+generateLastITableAndITableInstructions(TR::CodeGenerator * cg, TR::Node * callNode, TR::SymbolReference * methodSymRef, TR::Register * vftReg, TR::Register * scratchRegister,
+   TR::Register * loopCountRegister, TR::Register * vTableIndexRegister, TR::RegisterDependencyConditions * postDeps, TR::Instruction * cursor)
+   {
+   TR::Compilation * comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(comp->fe());
+   uintptr_t itableIndex;
+   TR_OpaqueClassBlock *declaringClass = NULL;
+   TR_Debug * debugObj = cg->getDebug();
+   #define iComment(str) if (debugObj) debugObj->addInstructionComment(cursor, (const_cast<char*>(str)));
+
+   static bool EnableLastITableCache = feGetEnv("TR_interfaceDispatchUsingLastITable") != NULL;
+
+   if ( EnableLastITableCache && (declaringClass = methodSymRef->getOwningMethod(comp)->getResolvedInterfaceMethod(methodSymRef->getCPIndex(), &itableIndex))
+      && performTransformation(comp, "O^O useLastITableCache for n%dn itableIndex=%d\n",
+            callNode->getGlobalIndex(), (int)itableIndex))
+      {
+      const int32_t DYNAMIC_LOOP_THRESHOLD = 2;
+      TR::LabelSymbol * noMatchLabel = generateLabelSymbol(cg);
+      TR::LabelSymbol * matchLabel = generateLabelSymbol(cg);
+
+      int32_t iTableIterations = getITableIterationsNumber(comp, methodSymRef, declaringClass);
+      // if the iteration is larger than 0x7fff, check all entries without counting.
+      bool checkAllITableEnrties = iTableIterations > INT16_MAX;
+      bool checkLimitedNumberOfITableEntries = !checkAllITableEnrties && iTableIterations > DYNAMIC_LOOP_THRESHOLD;
+      bool checkITableEntries = iTableIterations > 0;
+      bool stopUsingScratchRegister = false;
+
+      if (scratchRegister == NULL)
```

Comment:
We may have case where we will be allocating scratch register and one more register - Is it possible to use scratch register manager for this work, to simplify the code?